abstract/encoding.go

```diff
@@ -170,13 +193,31 @@ func (de *decoder) value(v reflect.Value, depth int) error {
 			}
 		}
 
-	case reflect.Slice, reflect.Array:
+	case reflect.Slice:
+		if v.IsNil() {
+			panic("slices must be initialized to correct length before decoding")
+		}
+		fallthrough
+	case reflect.Array:
 		l := v.Len()
 		for i := 0; i < l; i++ {
 			if err = de.value(v.Index(i), depth+1); err != nil {
 				return err
 			}
 		}
+
+	case reflect.Int:
```

Comment:
I'm not sure I quite understand what kinds of [heisen]bugs you're worried about here, or what you want to enforce about "what type of int to use" - can you clarify?  Avoiding silent heisenbugs is the reason I added the specific overflow check and error return if a 64-bit int arrives that's too large for a 32-bit platform to represent as an int internally.  Are there other important failure conditions I'm missing?